### PR TITLE
Add initial support for RISCV RVV features

### PR DESCRIPTION
**Pull request by arcbbb**
`mabi` is introduced because target triple doesn't encode ABI info,
and `HL_LLVM_ARGS="-riscv-v-vector-bits-min 128"` is required to enable fixed vectors to RVV codegen.

### CODE REVIEW

File: src/CodeGen_RISCV.cpp
Comment (steven-johnson) [inline src/CodeGen_RISCV.cpp:45]: Interesting, I didn't realize that RISC-V offered soft-float -- does this work? (I wouldn't be surprised if we have code somewhere that assumes it applies only to arm32, which AFAIK was this only place this was enabled previously)

File: src/CodeGen_Internal.cpp
Comment (steven-johnson) [reply src/CodeGen_Internal.cpp:695]: UPDATE: just realized that our buildbots are currently only testing PRs against LLVM 12 and 13, so this may not be easy to test. I'll work on fixing that.

**Review by steven-johnson — COMMENTED** [review]
LGTM so far, running against buildbots

File: src/CodeGen_Internal.cpp
Comment (steven-johnson) [reply src/CodeGen_Internal.cpp:695]: UPDATE #2: I tested locally against LLVM 10 and LLVM 11 and this seems to be fine as-is for those, so ignore this thread of comments.

**Comment by steven-johnson** [comment]
Two minor changes requested above, after which this is ready to land.

File: src/CodeGen_RISCV.cpp
Comment (arcbbb) [reply src/CodeGen_RISCV.cpp:24]: There is a logic to choose a default march in Clang & GCC,as described in [1].
When it is not explicitly specified, it is "rv32|64imafdc", which supports standard extensions:
- Integer Multiplication and Division,
- Atomic Instructions,
- Single-Precision Floating-Point, 
- Double-Precision Floating-Point,
- Compressed Instructions.

so I changed it to follow the convention.
[1] https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChains/Arch/RISCV.cpp#L721

File: src/CodeGen_RISCV.cpp
Comment (arcbbb) [reply src/CodeGen_RISCV.cpp:45]: I added it just in case someone want the generated library compatible with soft-float ABI.
It works fine as I tried `test/generator/example.generator` with `target=riscv-64-noos-rvv-soft_float_abi`.

File: src/CodeGen_Internal.cpp
Comment (arcbbb) [reply src/CodeGen_Internal.cpp:695]: Thanks!

File: src/CodeGen_RISCV.cpp
Comment (steven-johnson) [reply src/CodeGen_RISCV.cpp:25]: Ah, cool. It would probably be good to capture that in a comment in this function, since few readers will immediately know what those flags mean without reference.

**Comment by steven-johnson** [comment]
LGTM, pending adding the comments noted to CodeGen_RISCV::mattrs()

File: src/CodeGen_RISCV.cpp
Comment (arcbbb) [reply src/CodeGen_RISCV.cpp:31]: no problem!

**Review by steven-johnson — APPROVED** [review]
Thanks for the fix!

**Comment by dkurt** [comment]
Hi! Can you remember, which LLVM version was tested? Both LLVM 11 and LLVM 12 gives `llc: Unknown command line argument '-riscv-v-vector-bits-min'`.

**Comment by dkurt** [comment]
@arcbbb, I'm trying to find a combination of Halide+LLVM which will produce legal RVV code for AllWinner D1 (see full experiment [here](https://github.com/halide/Halide/discussions/7252)). I've tried this patch (commit https://github.com/halide/Halide/commit/5545e617aafc6815d3ebcf6a5c5f0a7f7c43311e) and LLVM revision https://github.com/llvm/llvm-project/commit/a719b667a9794ec0dc820d0c5a3fd18340521ad9. Despite they are close to each other by date, there is an exception when trying compile AOT:
```
app: /home/dkurt/llvm-project/llvm/include/llvm/CodeGen/MachineOperand.h:535: int64_t llvm::MachineOperand::getImm() const: Assertion `isImm() && "Wrong MachineOperand accessor"' failed.
```
```cpp
Func f("myfunc");
Buffer<uint8_t> input(width, height, 3);

Var x("x"), y("y");
f(x, y) = input(x, y, 0);
f.bound(x, 0, width).bound(y, 0, height);

f.vectorize(x, 8);

Target target;
target.os = Target::OS::Linux;
target.arch = Target::Arch::RISCV;
target.bits = 64;

std::vector<Target::Feature> features;
features.push_back(Target::RVV);
target.set_features(features);

f.compile_to_static_library("compiled", {input}, "myfunc", target);
```

**Comment by arcbbb** [comment]
> Hi! Can you remember, which LLVM version was tested? Both LLVM 11 and LLVM 12 gives `llc: Unknown command line argument '-riscv-v-vector-bits-min'`.
> 
> Probably, it's 13.x version which includes [llvm/llvm-project@a719b66](https://github.com/llvm/llvm-project/commit/a719b667a9794ec0dc820d0c5a3fd18340521ad9)

@dkurt I have a repo which records the version of LLVM & Halide I used at that time
Please check https://github.com/arcbbb/halide_testbed/tree/master/external

**Comment by dkurt** [comment]
@arcbbb, thank you! This revision still contains a problematic `vle8.v` but it's much closer to the state of LLVM where I can try revert https://github.com/llvm/llvm-project/commit/47a4a27f47203055a4700b65533262409f83c491.

**Pull request merged**
